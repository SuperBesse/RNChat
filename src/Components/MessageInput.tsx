import React, { useEffect, useRef } from 'react';
import { TextInput, StyleSheet, ActivityIndicator, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { sendMessage } from '@/Redux/Actions/MessageActions';
import { AppState } from '@/Redux/Reducers';
type Props = {};

const MessageInput = ({}: Props) => {
  const [text, setText] = React.useState<string>('');
  const textInputRef = useRef<TextInput>(null);
  //listen text changing but rerender only if text has changed
  const onChangeText = (value: string) => {
    if (value !== text) {
      setText(value);
    }
  };

  const { t } = useTranslation();
  const dispatch = useDispatch();

  //get status onf message to display loader and reset text on success
  const isSending = useSelector((state: AppState) => {
    const isLoading = state.messageState.isSending;
    if (isLoading) {
      onChangeText('');
    }
    return isLoading;
  });

  useEffect(() => {
    if (!isSending && textInputRef.current) {
      textInputRef.current.focus();
    }
  }, [isSending]);

  const sendNewMessage = () => {
    dispatch(sendMessage(text));
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.input}
        ref={textInputRef}
        onChangeText={onChangeText}
        value={text}
        placeholder={t('placeholder.message')}
        onEndEditing={sendNewMessage}
        editable={!isSending}
        enablesReturnKeyAutomatically
        autoCorrect={false}
        autoCapitalize={'none'}
      />
      {isSending && (
        <ActivityIndicator
          size="small"
          color="#grey"
          testID="activity-sending"
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  input: {
    alignSelf: 'center',
    borderColor: 'gray',
    width: '80%',
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 8,
    marginRight: 8,
  },
});

export default MessageInput;
