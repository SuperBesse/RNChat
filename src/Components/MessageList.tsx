import React from 'react';
import { FlatList, ViewStyle, ListRenderItem, StyleSheet } from 'react-native';
import Message from './Message';
import { Message as IMessage } from '@/Types/Message';
import { useSelector } from 'react-redux';
import { AppState } from '@/Redux/Reducers';

type Props = {
  style?: ViewStyle;
};

const MessageList = ({ style }: Props) => {
  const renderItem: ListRenderItem<IMessage> = ({ item }) => (
    <Message message={item} />
  );

  const messages = useSelector(
    (state: AppState) => state.messagesState.messages,
  );

  return (
    <FlatList
      style={[style, styles.list]}
      contentContainerStyle={style}
      data={messages}
      renderItem={renderItem}
      keyExtractor={(item: IMessage) => item.date.toString()}
      persistentScrollbar
      inverted
    />
  );
};

export default MessageList;

const styles = StyleSheet.create({
  list: {
    flex: 1,
  },
});
MessageList.defaultProps = {
  style: undefined,
};
