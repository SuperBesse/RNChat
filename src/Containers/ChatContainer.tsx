import React from 'react';

import { MessageInput, MessageList } from '@/Components';
import { StyleSheet, View } from 'react-native';

const ChatContainer = () => {
  return (
    <View style={styles.container}>
      <MessageList style={styles.container} />
      <MessageInput />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export default ChatContainer;
