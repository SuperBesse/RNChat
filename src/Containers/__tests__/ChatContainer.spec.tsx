import React from 'react';
import { render, screen } from '@testing-library/react-native';
import ChatContainer from '../ChatContainer';
import MessageInput from '@/Components/MessageInput';
import MessageList from '@/Components/MessageList';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe('ChatContainer', () => {
  it('should render correctly', () => {
    const tree = render(<ChatContainer />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(screen.UNSAFE_getAllByType(MessageInput).length).toBe(1);
    expect(screen.UNSAFE_getAllByType(MessageList).length).toBe(1);
  });
});
