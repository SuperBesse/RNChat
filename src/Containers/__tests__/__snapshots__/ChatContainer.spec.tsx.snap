// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatContainer should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    invertStickyHeaders={true}
    inverted={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    persistentScrollbar={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        },
        Array [
          undefined,
          Object {
            "flex": 1,
          },
        ],
      ]
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      onChangeText={[Function]}
      onEndEditing={[Function]}
      placeholder="placeholder.message"
      style={
        Object {
          "alignSelf": "center",
          "borderColor": "gray",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginRight": 8,
          "marginVertical": 8,
          "padding": 10,
          "width": "80%",
        }
      }
      value=""
    />
  </View>
</View>
`;
